influxdb3_write/src/write_buffer/buffer_segment.rs

```diff
@@ -0,0 +1,656 @@
+            self.segment_id,
+            self.starting_catalog_sequence_number,
+            catalog.sequence_number(),
+            self.segment_writer,
+            self.buffered_data,
+            catalog,
+        )
+    }
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct WriteBatch {
+    database_batches: HashMap<NamespaceName<'static>, DatabaseBatch>,
+}
+
+impl WriteBatch {
+    pub(crate) fn add_db_write(
+        &mut self,
+        db_name: NamespaceName<'static>,
+        table_batches: HashMap<String, TableBatch>,
+    ) {
+        let db_batch = self.database_batches.entry(db_name).or_default();
+        db_batch.add_table_batches(table_batches);
+    }
+}
+
+#[derive(Debug, Default)]
+struct DatabaseBatch {
+    table_batches: HashMap<String, TableBatch>,
+}
+
+impl DatabaseBatch {
+    fn add_table_batches(&mut self, table_batches: HashMap<String, TableBatch>) {
+        for (table_name, table_batch) in table_batches {
+            let write_table_batch = self.table_batches.entry(table_name).or_default();
+
+            for (partition_key, partition_batch) in table_batch.partition_batches {
+                let write_partition_batch = write_table_batch
+                    .partition_batches
+                    .entry(partition_key)
+                    .or_default();
+                write_partition_batch.rows.extend(partition_batch.rows);
+            }
+        }
+    }
+}
+
+pub struct BufferedWrite {
+    pub wal_op: WalOp,
+    pub database_write: DatabaseWrite,
+    pub response_tx: oneshot::Sender<BufferedWriteResult>,
+}
+
+pub struct DatabaseWrite {
+    pub(crate) db_name: NamespaceName<'static>,
+    pub(crate) table_batches: HashMap<String, TableBatch>,
+}
+
+impl DatabaseWrite {
+    pub fn new(
+        db_name: NamespaceName<'static>,
+        table_batches: HashMap<String, TableBatch>,
+    ) -> Self {
+        Self {
+            db_name,
+            table_batches,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct WriteSummary {
+    pub segment_id: SegmentId,
+    pub sequence_number: SequenceNumber,
+    pub buffer_size: usize,
+}
+
+#[derive(Debug, Default)]
+struct DatabaseBuffer {
+    table_buffers: HashMap<String, TableBuffer>,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct TableBuffer {
+    pub partition_buffers: HashMap<String, PartitionBuffer>,
+}
+
+impl TableBuffer {
+    #[allow(dead_code)]
+    pub fn partition_buffer(&self, partition_key: &str) -> Option<&PartitionBuffer> {
+        self.partition_buffers.get(partition_key)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct PartitionBuffer {
+    rows: Vec<Row>,
+    timestamp_min: i64,
+    timestamp_max: i64,
+}
+
+impl Default for PartitionBuffer {
+    fn default() -> Self {
+        Self {
+            rows: Vec::new(),
+            timestamp_min: i64::MAX,
+            timestamp_max: i64::MIN,
+        }
+    }
+}
+
+impl PartitionBuffer {
+    pub fn add_rows(&mut self, rows: Vec<Row>) {
```

Comment:
May be good to reserve prior to the loop:
```rust
self.rows.reserve(rows.len());
```
to reduce re-allocations in the loop when using `push`.